src/porepy/numerics/ad/_ad_utils.py

```diff
@@ -294,6 +294,42 @@ def discretize_from_list(
                     pass
 
 
+def _validate_indices(
```

Comment:
Adding general validation of time step and iteration indices.
Using ValueError to alert invalid usage (index increasing with depth)
Using KeyError to alert user that evaluation is attempted at index for which no data is set.